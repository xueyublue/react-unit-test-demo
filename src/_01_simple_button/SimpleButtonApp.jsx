import React, { useState } from "react";
import styles from "./SimpleButtonApp.module.css";

export default function SimpleButtonApp() {
  const [color, setColor] = useState("MediumVioletRed");
  const [disabled, setDisabled] = useState(false);
  const newColor = color === "MediumVioletRed" ? "MidnightBlue" : "MediumVioletRed";

  return (
    <div className={styles.container}>
      <button
        data-testid="colorButton"
        className={styles.button}
        style={{
          backgroundColor: disabled ? "gray" : color,
        }}
        onClick={() => setColor(newColor)}
        disabled={disabled}
      >
        Change to {replaceCamelWithSpaces(newColor)}
      </button>
      <input
        type="checkbox"
        id="disable-button-checkbox"
        defaultChecked={disabled}
        onChange={(e) => setDisabled(e.target.checked)}
      />
      <label htmlFor="disable-button-checkbox">Disable Button</label>
    </div>
  );
}

export function replaceCamelWithSpaces(colorName) {
  return colorName.replace(/\B([A-Z])\B/g, " $1");
}
