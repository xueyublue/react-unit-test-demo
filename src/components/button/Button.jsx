import React from "react";

export default function Button({ label }) {
  return <button data-testid="button">{label}</button>;
}
