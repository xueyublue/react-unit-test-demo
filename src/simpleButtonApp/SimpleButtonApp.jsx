import React, { useState } from "react";

export default function SimpleButtonApp() {
  const [color, setColor] = useState("red");
  const [disabled, setDisabled] = useState(false);
  const newColor = color === "red" ? "blue" : "red";

  return (
    <div style={{ width: "100vw", height: "100vh", display: "flex", alignItems: "center", justifyContent: "center" }}>
      <button
        data-testid="colorButton"
        style={{
          backgroundColor: disabled ? "gray" : color,
          color: "white",
          width: "25%",
          padding: "10px 10px",
          transition: "0.5s ease-in-out",
          border: "none",
          textTransform: "uppercase",
        }}
        onClick={() => setColor(newColor)}
        disabled={disabled}
      >
        Change to {newColor}
      </button>
      <input
        type="checkbox"
        id="disable-button-checkbox"
        defaultChecked={disabled}
        onChange={(e) => setDisabled(e.target.checked)}
      />
      <label htmlFor="disable-button-checkbox">Disable Button</label>
    </div>
  );
}
