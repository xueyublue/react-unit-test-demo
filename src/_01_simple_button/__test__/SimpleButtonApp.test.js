import { render, screen, cleanup, fireEvent, logRoles } from "@testing-library/react";
import Button from "../SimpleButtonApp";
import renderer from "react-test-renderer";
import SimpleButtonApp from "../SimpleButtonApp";
import { replaceCamelWithSpaces } from "..//SimpleButtonApp";

describe("test frontend behaviors", () => {
  beforeEach(cleanup);

  test("button has correct initial color", () => {
    const { container } = render(<Button />);
    logRoles(container);

    const colorButton = screen.getByRole("button", { name: /Change to midnight blue/i });
    expect(colorButton).toHaveStyle({ backgroundColor: "MediumVioletRed" });
  });

  test("button turns blue when clicked", () => {
    render(<Button />);
    const colorButton = screen.getByRole("button", { name: /Change to midnight blue/i });

    fireEvent.click(colorButton);

    expect(colorButton).toHaveStyle({ backgroundColor: "MidnightBlue" });
    expect(colorButton).toHaveTextContent(/Change to medium violet red/i);
  });

  test("matches snapshot", () => {
    const tree = renderer.create(<Button />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("initial conditions", () => {
    render(<Button />);

    const colorButton = screen.getByRole("button", /Change to midnight blue/i);
    expect(colorButton).toBeEnabled(colorButton);

    const checkbox = screen.getByRole("checkbox");
    expect(checkbox).not.toBeChecked(checkbox);
  });

  test("checkbox disables button on first click and enabls on second click", () => {
    render(<Button />);
    const checkbox = screen.getByRole("checkbox", { name: /disable button/i });
    const button = screen.getByRole("button", { name: /change to midnight blue/i });

    fireEvent.click(checkbox);
    expect(button).toBeDisabled();

    fireEvent.click(checkbox);
    expect(button).toBeEnabled();
  });

  test("Disable button has gray background and reverts to red", () => {
    render(<SimpleButtonApp />);
    const checkbox = screen.getByRole("checkbox", { name: /disable button/i });
    const button = screen.getByRole("button", { name: /change to midnight blue/i });

    fireEvent.click(checkbox);
    expect(button).toHaveStyle({ backgroundColor: "gray" });

    fireEvent.click(checkbox);
    expect(button).toHaveStyle({ backgroundColor: "MediumVioletRed" });
  });

  test("Disable button has gray background and reverts to blue", () => {
    render(<SimpleButtonApp />);
    const checkbox = screen.getByRole("checkbox", { name: /disable button/i });
    const button = screen.getByRole("button", { name: /change to midnight blue/i });

    // change to blue
    fireEvent.click(button);

    fireEvent.click(checkbox);
    expect(button).toHaveStyle({ backgroundColor: "gray" });

    fireEvent.click(checkbox);
    expect(button).toHaveStyle({ backgroundColor: "MidnightBlue" });
  });
});

describe("spaces before camel-case capital letters", () => {
  test("Works for no inner capital letters", () => {
    expect(replaceCamelWithSpaces("Red")).toBe("Red");
  });
  test("Works for one inner capital letter", () => {
    expect(replaceCamelWithSpaces("MidnightBlue")).toBe("Midnight Blue");
  });
  test("Works for multiple inner capital letters", () => {
    expect(replaceCamelWithSpaces("MediumVioletRed")).toBe("Medium Violet Red");
  });
});
