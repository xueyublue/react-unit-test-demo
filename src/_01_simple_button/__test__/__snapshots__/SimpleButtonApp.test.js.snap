// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test frontend behaviors matches snapshot 1`] = `
<div
  className="container"
>
  <button
    className="button"
    data-testid="colorButton"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "MediumVioletRed",
      }
    }
  >
    Change to 
    Midnight Blue
  </button>
  <input
    defaultChecked={false}
    id="disable-button-checkbox"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    htmlFor="disable-button-checkbox"
  >
    Disable Button
  </label>
</div>
`;
